Pegasus/include/pegasus/Collision.hpp

```diff
@@ -154,113 +245,148 @@ class Detector
                 }
             }
         }
-
-        return contacts;
     }
 };
 
 class Resolver
 {
 public:
-    uint32_t iterationsUsed;
-    uint32_t iterations = 10000;
+    Resolver() = default;
+
+    /**
+    * @brief Construct resolver initialized with a given asset manager
+    * @param assetManager scene's asset manager
+    */
+    Resolver(scene::AssetManager& assetManager);
 
     /**
      * @brief Resolves collisions
      * @param contacts contacts information
      * @param duration delta time of the frame
      */
-    void Resolve(std::vector<std::vector<Contact>>& contacts, double duration);
+    void Resolve(std::vector<Contact> contacts, double duration);
 
-private:
     /**
-     * @brief Calculates total separation speed of the contact
-     * @param contact contact information
-     * @return speed meters per second
+     * @brief Resolves cached contacts
      */
-    static double CalculateTotalSeparationSpeed(Contact contact);
+    void ResolvePersistantContacts(double duration); 
 
-    /**
-     * @brief Removes acceleration caused component from the separation speed
-     * @param contact contact information
-     * @param separationSpeed total separation speed
-     * @param duration delta time of the frame
-     * @return pure separation speed
-     */
-    static double CalculatePureSeparationSpeed(Contact contact, double totalSeparationSpeed, double duration);
-
-    /**
-     * @brief Calculates separation speed of the contact
-     * @param contact contact information
-     * @param duration delta time of the frame
-     * @return separation speed
-     */
-    static double CalculateSeparationSpeed(Contact contact, double duration);
+private:
+    double const m_persistentFactor = 0.05f;
+    double const m_persistentThreshold = 1e-3;
+    double const m_persistentThresholdSq = m_persistentThreshold * m_persistentThreshold;
+    scene::AssetManager* m_pAssetManager = nullptr;
+    std::vector<Contact> m_prevContacts;
+    std::vector<Contact> m_persistentContacts;
+
+    //!Calculates hashes for contacts based on handles
+    struct ContactHasher
+    {
+        size_t operator()(Contact const& c) const
+        {
+            return std::hash<scene::Handle>()(c.aBodyHandle)
+                ^ std::hash<scene::Handle>()(c.bBodyHandle);
+        }
 
+        bool operator()(Contact const& a, Contact const& b) const
+        {
+            return (a.aBodyHandle == b.aBodyHandle)
+                && (a.bBodyHandle == b.bBodyHandle);
+        }
+    };
+    
     /**
-     * @brief Calculates total impulse of the contact
-     * @param contact contact information
-     * @param duration delta time of the frame
-     * @return total impulse
+     * @brief Detects contacts existing during multiple frames
+     * @param contacts contact set for search
      */
-    static glm::dvec3 CalculateTotalImpulse(Contact contact, double duration);
+    void DetectPersistentContacts(
```

Comment:
It should not be called two times in a row, however, it is an internal method, so whoever changing it's call pattern better be understanding what he/she is doing. Documenting it would not make any change to the user code.

Reply:
It would at least explain the scope of changes by this method. Also it will help understanding the order of changes to underlying objects.